stdlib/Test/src/Test.jl

```diff
@@ -600,16 +600,28 @@ macro test_throws(extype, ex)
     :(do_test_throws($result, $orig_ex, $(esc(extype))))
 end
 
+const MACROEXPAND_LIKE = Symbol.(("@macroexpand", "@macroexpand1"))
+
 # An internal function, called by the code generated by @test_throws
 # to evaluate and catch the thrown exception - if it exists
 function do_test_throws(result::ExecutionResult, orig_expr, extype)
     if isa(result, Threw)
         # Check that the right type of exception was thrown
         success = false
         exc = result.exception
+        # NB: Throwing LoadError from macroexpands is deprecated, but in order to limit
+        # the breakage in package tests we add extra logic here.
+        from_macroexpand =
+            orig_expr isa Expr &&
+            orig_expr.head === :macrocall &&
+            orig_expr.args[1] in MACROEXPAND_LIKE
```

Comment:
Thanks for the review, and sorry it took me a while to get back to this...I'm a bit conflicted about where to go from here. Both `@macroexpand` and `macroexpand` could be called somewhere in the body of a complicated expression, or even from inside a separate function. Automatically passing `LoadError` without any further warning seems too dangerous though. In other words, I'm not sure it's possible to have a neat deprecation here?

Reply:
Well, to be honest I prefer the original version of this code from https://github.com/JuliaLang/julia/blob/3ee9d8f813dde0339a598524a484d4d822498101/stdlib/Test/src/Test.jl#L592

In that case I didn't try to detect where the `LoadError` came from, so the heuristic was very simple. Basically

"If within `@test_throws` and the user code specified `LoadError`, then it's ok for the actual thrown exception not to be a load error"

This is just to prevent a fair bit of spurious test breakage in actual practical cases where I know packages test their macro error messages using `@test_throws`.

Certainly any attempt at removing `LoadError` will be a minor breaking change for some type of end user code. But if we're a little over-permissive in allowing error types thrown within `@test_throws` at least the common cases will continue working and old package tests won't be spuriously broken in practice.

Reply:
But I'm not sure I fully understand your comment. I'm suggesting that being more permissive is ok here (I think it's ok just to remove the `from_macroexpand` check).

Reply:
Well, I guess I don't care too much either way, so happy to just remove the `from_macroexpand` check. An example where that might be a bit dangerous: perhaps a user could get confused between `LoadError` and `ArgumentError`, accidentally write `@test_throws LoadError something_that_throws_an_argument_error`, and continue on merrily when tests pass. Then, a bug in the user's code throws some other type of error, like `KeyError`. The bug wouldn't get caught.

Reply:
Fair comment. My thinking was that it's useful to provide a low-friction way to keep historical code working on both older Julia versions and the current one. Preventing users from writing new bugs is also a worthwhile aim.

If we don't detect calls to `macroexpand` (or just be generally more permissive as in the original PR) I know that various package tests will break in a spurious way. Examples can be found here https://github.com/search?q=test_throws+LoadError+macroexpand&type=code

One specific example:
https://github.com/GiggleLiu/NiLangCore.jl/blob/eaf26c9cde1fa30b37ed59d345dacda19eec77ff/test/compiler.jl#L104

Reply:
New bugs seem more serious that friction to me, but I can see how reducing both would be helpful. Maybe we could throw a a warning for `@test_throws [unqualified]LoadError [@]macroexpand[1] stuff`? Something like `@warn "macroexpands no longer throw a LoadError. Test passed without checking the error type!"`

Reply:
It depends on your perspective ;-) If you happen to be supporting a big codebase or a widely used library which is compatible across Julia versions, it's quite unfortunate to have changes to `Base` break your tests for all old versions.

> Maybe we could throw a a warning

Right, let's just call `Base.depwarn()`. That way, old versions of library tests will continue to work and the deprecation should be actively visible. In that case it also seems safe to remove the `from_macroexpand` check.